RemoteImageDecoderAVFProxy does not attribute the frames to the owner process

kkinnunen-apple opened the pull request:
#### 146c1b7c9159eb722750086e5c2691d5fa72da00
<pre>
RemoteImageDecoderAVFProxy does not attribute the frames to the owner process
<a href="https://bugs.webkit.org/show_bug.cgi?id=261877">https://bugs.webkit.org/show_bug.cgi?id=261877</a>
rdar://115838448

Reviewed by Brent Fulgham.

Attribute the IOSurface-backed CVPixelBuffers to the WebContent process
that instantiated the RemoteImageDecoderAVFProxy.

Note: this captures only subset of the CVPixelBuffers in flight.
Subsequent work is still needed.

* Source/WebCore/platform/graphics/ImageDecoder.cpp:
(WebCore::createInProcessImageDecoderAVFObjC):
(WebCore::platformRegisterFactories):
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::create):
(WebCore::ImageDecoderAVFObjC::ImageDecoderAVFObjC):
(WebCore::ImageDecoderAVFObjC::storeSampleBuffer):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::RemoteImageDecoderAVFProxy):
(WebKit::RemoteImageDecoderAVFProxy::createDecoder):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.h:

Canonical link: <a href="https://commits.webkit.org/268325@main">https://commits.webkit.org/268325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65dbf14755ad6db7838352a610032ea37169a046

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19343 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21232 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18098 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19886 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19560 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16818 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22101 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16792 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17606 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17854 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21902 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17506 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4625 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21866 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
```diff
@@ -424,6 +426,10 @@ SampleFlags flags() const override
         RELEASE_LOG_ERROR(Images, "ImageDecoderAVFObjC::storeSampleBuffer(%p) - could not decode sampleBuffer", this);
         return false;
     }
```
Nit: Should have a blank line before the comments start.

brentfulgham commented on Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
```diff
@@ -432,15 +438,22 @@ SampleFlags flags() const override
 
-    if (m_imageRotationSession)
+    if (m_imageRotationSession) {
         pixelBuffer = m_imageRotationSession->rotate(pixelBuffer.get());
```
If we are switching to the rotated version, should we do something to clear the Ownership Identity in the original pixel buffer?

kkinnunen-apple replied:
No, it still exists due to the requester requesting it. You can think of the size of the frame being 2x what you'd expect since it needs orientation change.
In this case the original will be destroyed as the ref goes out of scope, but until the destruction is real, the attribution (pedantically speaking) assigns the blame to the correct owner.

brentfulgham commented on Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
```diff
@@ -432,15 +438,22 @@ SampleFlags flags() const override
         return false;
     }
 
+    // FIXME: The pixel buffer being passed to the VTCreateCGImageFromCVPixelBuffer may
```
Nit: A FIXME should have an associated bugzilla ID in it so we don't forget to fix it.

brentfulgham reviewed (approved):
Looks good aside from a few small nits. r=me

kkinnunen-apple committed:
RemoteImageDecoderAVFProxy does not attribute the frames to the owner process
https://bugs.webkit.org/show_bug.cgi?id=261877
rdar://115838448

Reviewed by Brent Fulgham.

Attribute the IOSurface-backed CVPixelBuffers to the WebContent process
that instantiated the RemoteImageDecoderAVFProxy.

Note: this captures only subset of the CVPixelBuffers in flight.
Subsequent work is still needed.

* Source/WebCore/platform/graphics/ImageDecoder.cpp:
(WebCore::createInProcessImageDecoderAVFObjC):
(WebCore::platformRegisterFactories):
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::create):
(WebCore::ImageDecoderAVFObjC::ImageDecoderAVFObjC):
(WebCore::ImageDecoderAVFObjC::storeSampleBuffer):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::RemoteImageDecoderAVFProxy):
(WebKit::RemoteImageDecoderAVFProxy::createDecoder):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.h:

Canonical link: https://commits.webkit.org/268325@main
